### PR TITLE
fix(actions): swap sha256sum for shasum -a 256

### LINKED ISSUE #1389: Github Actions pipeline fails to generate sha256 sum for caching macOS builds
### Describe the problem

While using copier's ci pipeline as a template for use in another repo, I noticed that the `macOS-latest` builds were finishing consistently later than `ubuntu-latest` . Looking closely at the logs I noticed some errors in the cache key step. 

### Template

Not Relevant

### To Reproduce

https://github.com/copier-org/copier/actions/runs/6681655342/job/18156002683
View the step generate cache key PY

### Logs

```console
Run echo "PY=$((python -VV; pip freeze) | sha256sum | cut -d' ' -f1)" >> $GITHUB_ENV
  echo "PY=$((python -VV; pip freeze) | sha256sum | cut -d' ' -f1)" >> $GITHUB_ENV
  shell: /bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    LANG: en_US.utf-8
    LC_ALL: en_US.utf-8
    PIP_CACHE_DIR: /Users/runner/work/copier/copier/.cache/pip
    POETRY_CACHE_DIR: /Users/runner/work/copier/copier/.cache/pypoetry
    POETRY_VIRTUALENVS_IN_PROJECT: true
    PRE_COMMIT_HOME: /Users/runner/work/copier/copier/.cache/pre-commit
    PYTEST_ADDOPTS: 
    PYTHONIOENCODING: UTF-8
    pythonLocation: /Users/runner/hostedtoolcache/Python/3.12.0/x64
    PKG_CONFIG_PATH: /Users/runner/hostedtoolcache/Python/3.12.0/x64/lib/pkgconfig
    Python_ROOT_DIR: /Users/runner/hostedtoolcache/Python/3.12.0/x64
    Python2_ROOT_DIR: /Users/runner/hostedtoolcache/Python/3.12.0/x64
    Python3_ROOT_DIR: /Users/runner/hostedtoolcache/Python/3.12.0/x64
/Users/runner/work/_temp/6cc334fe-c655-429e-a1a9-e767bf611311.sh: line 1: sha256sum: command not found
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
BrokenPipeError: [Errno 32] Broken pipe
```

### Expected behavior

The sha256 key should be generated and used for caching.

### Screenshots/screencasts/logs

_No response_

### Operating system

macOS

### Operating system distribution and version

macOS-latest

### Copier version

N/A

### Python version

All Python Versions

### Installation method

local build

### Additional context

I fixed this problem in my repo by switching to `shasum -a 256` instead which works on both `ubuntu-latest` and `macOS-latest`

### PR DESCRIPTION
**Pull request by kenibrewer**
This addresses an issue in macOS builds may be impacting caching. The command `sha256sum` which doesn't exist in `macOS-latest` is replaced with `shasum -a 256` which is present in both `ubuntu-latest` and `macOS-latest`

Closes #1389

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -51,7 +51,7 @@ jobs:
       - name: generate cache key PY
         shell: bash
         run:
-          echo "PY=$((python -VV; pip freeze) | sha256sum | cut -d' ' -f1)" >>
+          echo "PY=$((python -VV; pip freeze) | shasum -a 256 | cut -d' ' -f1)" >>
```
Comment (sisp) [inline .github/workflows/ci.yml:54]:
```suggestion
          echo "PY=$((python -VV; pip freeze) | python -uc 'import hashlib, sys; print(hashlib.sha256(sys.stdin.buffer.read()).hexdigest())' | cut -d' ' -f1)" >>
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -141,7 +141,7 @@ jobs:
           python-version: "3.10"
       - name: generate cache key PY
         run:
-          echo "PY=$((python -VV; pip freeze) | sha256sum | cut -d' ' -f1)" >>
+          echo "PY=$((python -VV; pip freeze) | shasum -a 256 | cut -d' ' -f1)" >>
```
Comment (sisp) [inline .github/workflows/ci.yml:144]:
```suggestion
          echo "PY=$((python -VV; pip freeze) | python -uc 'import hashlib, sys; print(hashlib.sha256(sys.stdin.buffer.read()).hexdigest())' | cut -d' ' -f1)" >>
```

**Review by sisp — CHANGES_REQUESTED** [review]
Thanks for discovering this error and taking the initiative to fix it, @kenibrewer! :pray:

Your current fix works for macOS and Ubuntu, but it doesn't work for Windows:

```
D:\a\_temp\29fa6760-37f4-4d7b-9324-e184f44d2852.sh: line 1: shasum: command not found
```

I see three options:

1. Use Python's `hashlib` module (see my inline suggestions).
1. Use the [cache key computation from Poetry's own CI config](https://github.com/python-poetry/poetry/blob/e12ca037bf0ff5faa93179270ef7c012f8e9ecdd/.github/workflows/main.yml#L76-L90).
1. Make this step conditioned on the OS and add another one with `sha256sum` for Windows.

**Comment by kenibrewer** [comment]
> 1. Use Python's `hashlib` module (see my inline suggestions).
> 2. Use the [cache key computation from Poetry's own CI config](https://github.com/python-poetry/poetry/blob/e12ca037bf0ff5faa93179270ef7c012f8e9ecdd/.github/workflows/main.yml#L76-L90).
> 3. Make this step conditioned on the OS and add another one with `sha256sum` for Windows.
> 
> I'm leaning towards solution 2, but solution 1 is less intrusive. Solution 3 adds too much duplicate code for my taste, but I wanted to mention it nevertheless.

I like solution 2 best as well. It's a lot more readable while solution 1 is pretty hard to parse.

**Comment by kenibrewer** [comment]
@sisp Can I get another workflow run?

**Comment by kenibrewer** [comment]
Looks like that didn't work for the windows install either. 

```
Run echo version=$(python -c "import sys; print('-'.join(str(v) for v in sys.version_info))") >> $GITHUB_OUTPUT
  echo version=$(python -c "import sys; print('-'.join(str(v) for v in sys.version_info))") >> $GITHUB_OUTPUT
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
  env:
    LANG: en_US.utf-8
    LC_ALL: en_US.utf-8
    PIP_CACHE_DIR: D:\a\copier\copier/.cache/pip
    POETRY_CACHE_DIR: D:\a\copier\copier/.cache/pypoetry
    POETRY_VIRTUALENVS_IN_PROJECT: true
    PRE_COMMIT_HOME: D:\a\copier\copier/.cache/pre-commit
    PYTEST_ADDOPTS: 
    PYTHONIOENCODING: UTF-8
    pythonLocation: C:\hostedtoolcache\windows\Python\3.9.13\x64
    PKG_CONFIG_PATH: C:\hostedtoolcache\windows\Python\3.9.13\x64/lib/pkgconfig
    Python_ROOT_DIR: C:\hostedtoolcache\windows\Python\3.9.13\x64
    Python[2](https://github.com/copier-org/copier/actions/runs/6683900995/job/18160692065?pr=1390#step:5:2)_ROOT_DIR: C:\hostedtoolcache\windows\Python\[3](https://github.com/copier-org/copier/actions/runs/6683900995/job/18160692065?pr=1390#step:5:3).9.13\x6[4](https://github.com/copier-org/copier/actions/runs/6683900995/job/18160692065?pr=1390#step:5:4)
    Python3_ROOT_DIR: C:\hostedtoolcache\windows\Python\3.9.13\x64

Cache restored successfully
Cache restored from key: cache|Windows||212c846ea7c96894e2ee344b3ed706190f898f3da49de7092b13c63149eea81b|c860f4[5]
```

I set up a test PR into my own fork so I can run the workflows myself. I'll probably test option 1 next and see if I can make it clearer and readable.

**Comment by kenibrewer** [comment]
@sisp This is now passing all tests and generates cache keys on all platforms.

**Review by yajo — APPROVED** [review]

**Review by sisp — APPROVED** [review]
Perfect, @kenibrewer, thank you! :pray:
